lib/recursive-methods.rb

```diff
@@ -1,49 +1,140 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    if n < 0
+        raise ArgumentError, "Will only accept positive intergar"
+    end
+
+    return 1 if  n == 0 
+    
+    return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    return s if s.length == 0
+
+    return "#{s[-1]}#{reverse(s[0...(s.length-1)])}"
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+
+    def reverse_helper( s, first, last = s.length - 1 )
+        if first < last
+            temp = s[first]
+            s[first] = s[last]
+            s[last] = temp
+            ## s[first] , s[last] = s[last], s[first]
+            return reverse_helper(s, first + 1, last - 1)
+        end
+
+        return s
+    end
+
+  return reverse_helper( s, 0, s.length - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+    if n < 0
+        raise ArgumentError, "can't have negative number for the bunny count!"
+    end
+    return 0 if n == 0
+
+    return 2 + bunny(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def nested(s)
```

Comment:
Since you are making a new string with each recursive call this is O(n^2) space complexity.